src/Lock.php

```diff
@@ -50,18 +55,24 @@ public function getId(): int
 
     /**
      * Releases the lock. No-op if the lock has already been released.
+     *
+     * @return Promise<void>
      */
-    public function release()
+    public function release(): Promise
     {
-        if (!$this->releaser) {
-            return;
+        if ($this->promise) {
+            return $this->promise;
         }
 
         // Invoke the releaser function given to us by the synchronization source
         // to release the lock.
-        $releaser = $this->releaser;
-        $this->releaser = null;
-        $releaser($this);
+        $promise = ($this->releaser)($this->id);
+        if (! $promise instanceof Promise) {
+            throw new InvalidArgumentException('The releaser must return a Promise.');
```

Comment:
Should the releaser be required to return a Promise?

Reply:
Not sure, we could also use `Amp\call` on it to automatically run as coroutine if it returns a generator.

Reply:
Nah, I'd avoid using call. I think the API should be more strict than that. I'm asking only because allowing void as well (which call wouldn't solve btw.) would be more BC friendly since that's how the releasers work now. Personally I'd simply force a Promise though.

Reply:
We use `Amp\call` on most other callbacks. It would seem unusual not to do so here as well IMO.

Reply:
I see. Okay then.